### PR TITLE
✅ [Idea]: KMP Algorithm for Pattern Searching

### LINKED ISSUE #1343: [Idea]: KMP Algorithm for Pattern Searching
### Idea Title

KMP Algorithm for Pattern Searching

### Idea Description

Let suppose we have been given two strings txt and pat of size N and M, where N > M. String txt and pat represent the text and pattern respectively. The task is to print all indexes of occurrences of pattern string in the text string. Use one-based indexing while returning the indices.

### Potential Benefits

- [ ] Improves user experience
- [X] Enhances learning
- [X] Adds a new feature
- [ ] Improves code quality
- [ ] Other

### Implementation Suggestions (Optional)

I wanted to add algorithms in the different programming languages.

### PR DESCRIPTION
**Pull request by yashksaini-coder**
### Description

This pull request adds comprehensive documentation for the KMP (Knuth-Morris-Pratt) pattern searching algorithm. The changes include creating a new category for the KMP algorithm and adding a detailed README file that explains the algorithm, its implementation, and usage examples.

Documentation additions:

* Created a new category for the KMP searching pattern with a description of the algorithm.
*  Added a detailed README file that includes an overview of the KMP algorithm, key features, how it works, implementation in Python, example usage, step-by-step example, performance comparison, common pitfalls, additional resources, and best practices.

✅ Fixes #1343 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by yashksaini-coder** [comment]
@ajay-dhangar can you review this PR, and assign it  to me with proper labels

**Comment by yashksaini-coder** [comment]
@ajay-dhangar can you review this PR and add proper labels to it.

**Comment by yashksaini-coder** [comment]
@ajay-dhangar can you add `hacktoberfest` labels too

**Comment by yashksaini-coder** [comment]
@ajay-dhangar I closed the PR to resolve the merge conflicts, cleared the conflicts, you can review it now.

File: docs/KMP searching pattern/readme.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
+# 🔍 KMP (Knuth-Morris-Pratt) Pattern Searching Algorithm
+
```
Comment (ajay-dhangar) [inline docs/KMP searching pattern/readme.md:10]: write code based on our format

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on our format and syntax

File: docs/KMP searching pattern/readme.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+---
+
+
+# 🔍 KMP (Knuth-Morris-Pratt) Pattern Searching Algorithm
+
+## 📚 Overview
+The KMP algorithm is an efficient string matching algorithm that finds occurrences of a "word" W within a main "text string" S by employing the observation that when a mismatch occurs, the word itself embodies sufficient information to determine where the next match could begin, thus bypassing re-examination of previously matched characters.
+
+<div align="center">
+    <img src="https://media.geeksforgeeks.org/wp-content/uploads/20221125004358/image-660x398.png" alt="KMP Algorithm" width="500">
```
Comment (ajay-dhangar) [inline docs/KMP searching pattern/readme.md]: is a non-void element. It requires a closing tag. Replace `` with `` to make it self-closing.

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
Please follow the `.jsx` and `.mdx` syntax for this pull request. Our project uses markdown plus React, so this syntax is required.

**Comment by yashksaini-coder** [comment]
@ajay-dhangar can you review this PR, and add hacktoberfest labels too

**Comment by yashksaini-coder** [comment]
@ajay-dhangar I have updated the code based on your specific requirements

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
Please recheck the code in your `docs/KMP searching pattern/readme.md` file. The content is not good, and does not meet our guidelines. Please revise the content to be more meaningful.

**Comment by yashksaini-coder** [comment]
> Please recheck the code in your `docs/KMP searching pattern/readme.md` file. The content is not good, and does not meet our guidelines. Please revise the content to be more meaningful.

I'm sorry for that, as it happens I had many PRs in your repo so am having problems in handling all, No worries I will update it later tomorrow.

Also Happy Diwali 🎆 @ajay-dhangar

**Comment by yashksaini-coder** [comment]
@ajay-dhangar I have added implementation in different languages with step-by-step explanation, Hope this changes are good

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**